### PR TITLE
Add link into the error snackbar

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds a link to the error messages. This PR is part of the changes described in https://github.com/woocommerce/woocommerce/issues/46768

Screenshot 2024-06-10 at 10 49 19 AM

### ACs
- [ ] If there is a field with an error in a non-visible tab, the error snackbar won't dismiss automatically and it will have a link to the error and a close button (`x`).
- [ ] Clicking the error link should navigate to the tab containing the field with an error and focus on the field. Currently, this works for: product name, regular price, and sale price.
- [ ] If there is a field with an error in a visible tab, the error snackbar won't include a close button (`x`) and will dismiss automatically. 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure `Try the new product editor (Beta)` is enabled at `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`.
2. Go to Products > Add New.
3. Add a regular price (but do not add a product name), and go to the Inventory tab.
4. Click Publish.
5. A snackbar with the message: `Product name is required. View error` should appear.
6. Press `View error`. You should be redirected to the General tab with the name field focused.
7. Repeat the test with a negative regular price or a negative sale price. The behavior should be the same.
8. Go to the Variations tab and create a few variations.
9. Select a single variation, navigate to the Shipping tab and toggle the `This variation requires shipping or pickup`.
10. Press `Update`.
11. A snackbar with the message: `Regular price is required. View error` should appear.
12. Press `View error`. You should be redirected to the General tab with the regular price field focused.
13. Add a price and switch to the Inventory tab.
14. Add a duplicated SKU.
15. Return to the General tab and click Update.
16. A snackbar with the message: `Invalid or duplicated SKU. View error` should appear.
17. Press `View error`. You should be redirected to the Inventory tab. Note that this error type is not fully handled yet.
18. Same with the dimensions checks under the Shipping tab. This will be fixed after merging [this PR](https://github.com/woocommerce/woocommerce/pull/49357).

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx [hunk 1]
```diff
@@ -75,6 +75,12 @@ export function ValidationProvider< T >( {
 		return Promise.resolve( undefined );
 	}
 
+	async function getFieldByValidatorId(
+		_validatorId: string
```
Comment (nathanss) [inline packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx]: I don't see a reason to use an underline in the name of the variable here

File: packages/js/product-editor/src/hooks/use-error-handler.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+function getErrorPropsWithActions(
+	errorContext = '',
+	validatorId: string,
+	focusByValidatorId: ( validatorId: string ) => void
+): ErrorProps {
+	return {
+		explicitDismiss: true,
+		actions: [
+			{
+				label: 'View error',
```
Comment (nathanss) [inline packages/js/product-editor/src/hooks/use-error-handler.ts]: Missing translation here

File: packages/js/product-editor/src/utils/test/get-product-error-message.ts [hunk 3]
```diff
@@ -1,46 +0,0 @@
-/**
- * Internal dependencies
- */
-import {
-	getProductErrorMessageAndProps,
-	WPError,
-} from '../get-product-error-message-and-props';
-
-describe( 'getProductErrorMessageAndProps.message', () => {
```
Comment (nathanss) [inline packages/js/product-editor/src/utils/test/get-product-error-message.ts:9]: You're deleting these tests but the production code was move elsewhere, right? Shouldn't the tests be adapted instead?

**Review by nathanss — CHANGES_REQUESTED** [review]
Thanks for implementing this! I tested and it's working as expected.

Left some feedback and questions regarding the code.

File: packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx (hunk 1 above)
Comment (octaedro) [reply packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx]: Done in commit [377950d](https://github.com/woocommerce/woocommerce/pull/49246/commits/377950d493c00c187cabcece1f9575a6b621f337)

File: packages/js/product-editor/src/hooks/use-error-handler.ts (hunk 2 above)
Comment (octaedro) [reply packages/js/product-editor/src/hooks/use-error-handler.ts]: Done in commit [377950d](https://github.com/woocommerce/woocommerce/pull/49246/commits/377950d493c00c187cabcece1f9575a6b621f337)

File: packages/js/product-editor/src/utils/test/get-product-error-message.ts (hunk 3 above)
Comment (octaedro) [reply packages/js/product-editor/src/utils/test/get-product-error-message.ts:9]: In order to make this PR smaller, I created a [separate PR](https://github.com/woocommerce/woocommerce/pull/49498) to tackle that. It's still in progress.

**Comment by octaedro** [comment]
Hey @nathanss, I just addressed your suggestions.
Could you take another look at this PR?

**Review by nathanss — APPROVED** [review]
Looking good! Approved.

**Pull request merged**